judge/tasks/user.py

```diff
@@ -17,19 +18,18 @@
 def apply_submission_filter(queryset, options):
     if not options['submission_download']:
         return []
+
+    use_straight_join(queryset)
+
     if options['submission_results']:
         queryset = queryset.filter(result__in=options['submission_results'])
 
-    problem_code_glob_cache = {}
-
-    def problem_code_glob_match(obj):
-        code = obj.problem.code
-        result = problem_code_glob_cache.get(code)
-        if result is None:
-            result = problem_code_glob_cache[code] = fnmatch.fnmatch(code, options['submission_problem_glob'])
-        return result
+    if options['submission_problem_glob'] != '*':
+        queryset = queryset.filter(
+            problem__in=Problem.objects.filter(code__regex=fnmatch.translate(options['submission_problem_glob'])),
```

Comment:
`fnmatch.translate` seems to be vulnerable to the generation of pathological expressions:

```pycon
>>> import fnmatch, re
>>> re.match(fnmatch.translate('*'*100+'b'), 'a'*1000)
<does not finish in a reasonable timeframe>
```

However, the query
```sql
SELECT 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa' REGEXP '(?s:.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*.*c)\\Z';
```
does seem to complete quickly. I guess this is okay?

Reply:
On a second thought it's probably best if we did `s/\*+/*/g` on the glob pattern.